src/Database/LSMTree/Internal/RawPage.hs

```diff
@@ -0,0 +1,63 @@
+{-# LANGUAGE RecordWildCards #-}
+module Database.LSMTree.Internal.RawPage (
+    RawPage (..),
+    rawPageNumKeys,
+    rawPageNumBlobs,
+    -- * Debug
+    rawPageKeyOffsets,
+) where
+
+import           Data.Bits (Bits, unsafeShiftR)
+import           Data.Primitive.ByteArray (ByteArray (..), indexByteArray)
+import           Data.Vector.Primitive (Vector (..))
+import           Data.Word (Word16)
+
+-------------------------------------------------------------------------------
+-- RawPage type
+-------------------------------------------------------------------------------
+
+newtype RawPage = RawPage ByteArray
```

Comment:
This is all very nice, and I like using ByteArray and the primitive indexing.

However, we have a constraint arising from the I/O: to get maximum IOPS we need to open the page files with `O_DIRECT`, and use direct I/O. This means the memory buffers need to be aligned to 4k. Last time I checked, it's pretty wasteful to use aligned bytearray allocation with 4k alignment, when allocating a 4k bytearray: it uses two 4k pages due to the heap object overhead.

So one idea is that we allocate buffers in bigger chunks, e.g. 10x 4k pages in one go, and thereby amortise the overhead. If we do that, then our `RawPage` type would need a `ByteArray` and offset (or direct pointer/addr).